platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

```diff
@@ -0,0 +1,493 @@
+ * to manage node participation in the Threshold Signature Scheme (TSS).<br/>
+ * All fields are REQUIRED.
+ */
+message RosterEntry {
+
+    /**
+     * A consensus node identifier.
+     * <p>
+     * Node identifiers SHALL be unique _within_ a shard and realm,
+     * but a node SHALL NOT, ever, serve multiple shards or realms,
+     * therefore the node identifier MAY be repeated _between_ shards and realms.
+     */
+    uint64 node_id = 1;
+
+    /**
+     * A consensus weight.
+     * <p>
+     * Each node SHALL have a weight of zero or more in consensus calculations.<br/>
+     * The consensus weight of a node SHALL be calculated based on the amount
+     * of HBAR staked to that node.<br/>
+     * Consensus SHALL be calculated based on agreement of greater than `2/3`
+     * of the total `weight` value of all nodes on the network.
+     */
+    uint64 weight = 2;
+
+    /**
+     * An RSA public certificate used for signing gossip events.
+     * <p>
+     * This value SHALL be a certificate of a type permitted for gossip
+     * signatures.<br/>
+     * This value SHALL be the DER encoding of the certificate presented.<br/>
+     * This field is REQUIRED and MUST NOT be empty.
+     */
+    bytes gossip_ca_certificate = 3;
+
+    /**
+     * An elliptic curve public signing key.<br/>
+     * This contains the _long term_ public key for this node.
+     * <p>
+     * This value SHALL be the DER encoding of the presented elliptic curve
+     * public key.<br/>
+     * This field is OPTIONAL (that is, it can initially be null)
+     * but once set, it can no longer be null.
+     */
+    bytes tss_encryption_key = 4;
+
+    /**
+     * A list of service endpoints for gossip.
+     * <p>
+     * These endpoints SHALL represent the published endpoints to which other
+     * consensus nodes may _gossip_ transactions.<br/>
+     * If the network configuration value `gossipFqdnRestricted` is set, then
+     * all endpoints in this list SHALL supply only IP address.<br/>
+     * If the network configuration value `gossipFqdnRestricted` is _not_ set,
+     * then endpoints in this list MAY supply either IP address or FQDN, but
+     * SHALL NOT supply both values for the same endpoint.<br/>
+     * This list SHALL NOT be empty.<br/>
+     * This list SHALL NOT contain more than `10` entries.<br/>
+     * The first two entries in this list SHALL be the endpoints published to
+     * all consensus nodes.<br/>
+     * All other entries SHALL be reserved for future use.
+     */
+    repeated proto.ServiceEndpoint gossip_endpoint = 5;
+}
+```
+
+#### Roster Map
+
+A States API map `Roster Map` (separate from the PBJ Map modeled above) will be introduced.
+It will contain keys which will be the hash of the roster (of type `byte`) and values of type `Roster`.
+This map will store all candidate and active rosters.
+It will be an implementation of `ReadableKVState<>` and `WriteableKVState<>` interfaces.
```

Comment:
It's very unlikely this will be a new implementation of these interfaces; it would be an instance of an `OnDiskWritableKVState`.

I think the indirection here makes things a lot less explicit, however. I would suggest instead three `Singleton` states:
1. _(Always set)_ A `RoundRosterPair` with the current active roster and the round it became active.
2. _(Sometimes set)_ A `RoundRosterPair` with the previous active roster and the  round _it_ became active.
3. _(Sometimes set)_ A `Roster` with the candidate roster.

Reply:
Ah, I've just looked again. Agreed on the instance of OnDiskWritableKVState.

We do need a `Roster` store that is forward-looking, considering we have Dynamic Address Book (DAB) on the horizon. In DAB, there may be many previous active `Roster`s we must keep track of. The use of singletons for storing the `Roster` today means we introduce a lot more state-change/data structure work to get to DAB. I'm not sure that's worth it. @jasperpotts, thoughts?

Reply:
I am torn on this. One one had, i don't think we should limit the number of rosters we can store in state by creating specific singletons for each. In fully dynamic address book, we need to track the active roster for every non-ancient round. It's true that even with fully dynamic address book, it's hard to image we will ever have more than the three rosters in state at one time especially given that it takes time to adopt a candidate roster, and 26 rounds is about 26 seconds today. But if things change in the future, like increasing the number of rounds non-ancient or decreasing the time for adoption of a new roster - we would need to change the state to support more rosters.

On the other hand, we know that we want to _decrease_ the number of rounds non-ancient in the future, the chance of ever needing more than these three is very small, and making them explicit is much clearer.

I do not think that storing them as a list in a singleton is the best option because that will increase the size of the block stream.